src/odm_sharing/private/queries.py

```diff
@@ -1,14 +1,326 @@
+        op_str = n.str_val.upper()
+        assert op_str in ['AND', 'OR']
+        arg_start = len(args)
+        sql = seq(n.sons)\
+            .map(recurse)\
+            .reduce(lambda x, y: f'({x} {op_str} {y})')
+        record(n, sql, args[arg_start:])
+        return sql
+    elif n.kind == NodeKind.FILTER:
+        # filter has op as value, children define field, kind and literals
+        op = parse_op(n.str_val)
+        key_ident = ident(recurse(n.sons[0]))
+
+        def gen_range_sql(range_kind: RangeKind, values: List[str]) -> str:
+            '''generates sql for a range of values'''
+            if range_kind == RangeKind.INTERVAL:
+                a = values[0]
+                b = values[1]
+                return f'({key_ident} BETWEEN {a} AND {b})'
+            elif range_kind == RangeKind.SET:
+                return f"({key_ident} IN ({','.join(values)}))"
+            else:
+                assert False, "unreachable"
+                return ''
+
+        if op == Op.RANGE:
+            range_kind = RangeKind(n.sons[1].str_val)
+            assert range_kind != RangeKind.SINGLE
+            literals: List[str] = seq(n.sons[2:]).map(recurse).list()
+            sql = gen_range_sql(range_kind, literals)
+            record(n, sql, args[-len(literals):])
+            return sql
+        else:
+            val = recurse(n.sons[1])
+            sql = f'({key_ident} {op} {val})'
+            record(n, sql, args[-1:])
+            return sql
+    elif n.kind == NodeKind.FIELD:
+        # field name is quoted
+        return dqt(n.str_val)
+    elif n.kind == NodeKind.LITERAL:
+        # literal value is converted and added to list of args, while only a
+        # parameter placeholder is added to the sql
+        val = convert(n.str_val)
+        args.append(val)
+        return '?'
+    else:
+        assert False, 'unreachable'
+        return ''
+
+
+def gen_sql(
+    root_node: Node,
+    rule_queries: Dict[RuleId, PartialQuery],
+) -> Query:
+    '''generates sql from a node-tree
+
+    :param root_node: the root-node of the tree to generate sql from
+    :param rule_queries: (output) partial filter and select queries
+
+    :return: complete query
+    '''
+    args: List[str] = []
+    sql = gen_sql_aux(root_node, args, rule_queries)
+    return Query(sql=sql, args=args)
+
+
+def get_table_node_columns(table_node: Node) -> List[str]:
+    '''returns the column-values of a select-node of a table-node'''
+    assert table_node.kind == NodeKind.TABLE
+    select_node = table_node.sons[0]
+    assert select_node.kind == NodeKind.SELECT
+    return seq(select_node.sons).map(lambda x: x.str_val).list()
+
+
+def parse_sql_dialect(dialect_str: str) -> SqlDialect:
+    '''parse dialect str to enum value, or fall back to default'''
+    try:
+        return SqlDialect(dialect_str.lower())
+    except ValueError:
+        return SqlDialect.OTHER
+
+
+def get_share_query(
+    table_node: Node,
```

Comment:
for each table in a share-node i need to find a matching query so that I can show the final count that share-rule would give for that table, if that makes sense?

Reply:
Can you update the docstring saying that it returns the query for a single table in a share node. The doc string right seems to imply that it returns the query for an entire share node.

Reply:
done

Reply:
done